spswarehouse/powerschool/powerschool.py

```diff
@@ -0,0 +1,437 @@
+    recent_file = recent_file.replace('\\', '/')
+    new_file, file_ext = os.path.splitext(recent_file)
+    new_file += append_text
+    new_file += file_ext
+    os.rename(recent_file, new_file)
+
+def download_latest_report_from_report_queue_reportworks(driver: WebDriver, destination_directory_path: str = '', file_postfix: str = ''):
+    """
+    Navigates to the PowerSchool Report Queue (ReportWorks), confirms the most recent report is done generating, and downloads it.
+
+    Parameters:
+    driver: A Selenium WebDriver
+    destination_directory_path: Where to download the PowerSchool report to.
+    file_postfix: Optional postfix to attach to the end of the downloaded file's filename.
+
+    Returns:
+    bool: True once successfully downloads the report. Otherwise, function keeps looping.
+    """
+
+    # Pause briefly to give a just-submitted report time to get into the queue
+    time.sleep(1)
+
+    ensure_on_desired_path(driver, REPORT_QUEUE_REPORTWORKS_PAGE_PATH)
+    
+    while True:
+        # TODO: Add a counter so this function can't get stuck in an infinte loop.
+        try:
+            # Confirm no reports are running
+            driver.find_element(By.XPATH, "//p[contains(text(), 'No reports running or pending!')]")
+            
+            # There is occasional flakiness where the "No reports running or pending!" message shows up but the 
+            # latest report is not in the list for downloading yet, so refresh the page one more time.
+            time.sleep(1)
+            WebDriverWait(driver, 5).until(EC.element_to_be_clickable((By.ID, 'prReloadButton')))
+            driver.find_element(By.ID, 'prReloadButton').click()
+            time.sleep(1)
+        except NoSuchElementException: # Because reports ARE running
+            time.sleep(5)
+            WebDriverWait(driver, 5).until(EC.element_to_be_clickable((By.ID, 'prReloadButton')))
+            driver.find_element(By.ID, 'prReloadButton').click()
+            logging.info('PowerSchool report is not ready, refreshing and waiting.')
+            time.sleep(3)
+        else:
+            # Download the first report in table
+            queued_reports = WebDriverWait(driver, 30).until(EC.element_to_be_clickable((By.XPATH, '//*[@id="queuecontent"]/table/tbody/tr[2]/td[7]/a')))
+            download_link = queued_reports.get_attribute('href')
+            original_files_list = os.listdir(destination_directory_path)
+            driver.get(download_link) #downloads the file
+            logging.info('PowerSchool report downloaded.')
+            break
+
+    wait_for_new_file_in_folder(destination_directory_path, original_files_list)
+    rename_recent_file_in_dir(destination_directory_path, file_postfix)
+    logging.info('Successfully renamed the downloaded file.')
+
+    return True
+
+def download_latest_report_from_report_queue_system(driver: WebDriver, destination_directory_path: str = '', file_postfix: str = ''):
```

Comment:
Also seems like an internal function?

Reply:
Hmm, ok, after finishing looking at this file, maybe this isn't an internal function. This does seem like the actual usable function in the end.

Reply:
I left this as external.